### ISSUE TITLE
Send_Progress_Notification Handler client_id getting set to nil

### ISSUE THREAD

**Issue by thecontinium** [jose-elias-alvarez/null-ls.nvim#1292]
### FAQ

- [X] I have checked the [FAQ](https://github.com/jose-elias-alvarez/null-ls.nvim#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Neovim Version

NVIM v0.9.0-dev-1715+ga6b05cb75-dirty

### Dev Version?

- [ ] I am using a stable Neovim release version, or if I am using a dev version of Neovim I have confirmed that my issue is reproducible on a stable version.

### Operating System

MacOS

### Minimal Config

```Lua
I am not adding a config here because it is a raise condition on exit of Neovim that is down to timing and does not replicate easily. Please see implementation help.
```

### Steps to Reproduce

Quiting neovim under certain circumstances which are hard to reproduce.

### Reproducibility Check

- [ ] I confirm that my minimal config is based on the `minimal_init.lua` template and that my issue is reproducible by running `nvim --clean -u minimal_init.lua` and following the steps above.

### Expected Behavior

No error

### Actual Behavior

The following error:

``` sh
Error executing vim.schedule lua callback: ...HEAD-a6b05cb/share/nvim/runtime/lua/vim/lsp/handlers.lua:29: bad argument #2 to 'format' (number expected, got nil)
stack traceback:
        [C]: in function 'format'
        ...HEAD-a6b05cb/share/nvim/runtime/lua/vim/lsp/handlers.lua:29: in function 'handler'
        ...m/jose-elias-alvarez/null-ls.nvim/lua/null-ls/client.lua:232: in function <...m/jose-elias-alvarez/null-ls.nvim/lua/null-ls/client.lua:231>
```

### Debug Log

See implementation help

### Help

No

### Implementation Help

When Neovim exits the following error occurs `Error executing vim.schedule lua callback: ...HEAD-a6b05cb/share/nvim/runtime/lua/vim/lsp/handlers.lua:29: bad argument #2 to 'format' (number expected, got nil)`

Here is the first few lines of the code in Neovim handlers.lua ( the last line shown here is where the format function fails ) :

``` lua
---@private
local function progress_handler(_, result, ctx, _)
  print (vim.inspect(result))
  print (vim.inspect(ctx))
  local client_id = ctx.client_id
  local client = vim.lsp.get_client_by_id(client_id)
  local client_name = client and client.name or string.format('id=%d', client_id)

....
```

This is happening because the client_id set [here ](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L238) is nil.

The erroneous call is passing:

``` lua

result = {
  token = 6,
  value = {
    kind = "end",
    percentage = 100
  }
}

ctx = {}
```

I'm not too familiar with Lua but on investigation I see the [on_exit](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L63) function is getting called, setting [id = nil](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L68), after this [test for nil](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L219) in the [send_progress_notification](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L218) function and ends up sending [client_id](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/623cc25016647eb62392aead7612f27d539c33de/lua/null-ls/client.lua#L238) = nil which causes the error.

I proved this by changing the on_exit function to not set `id` to nil like this:

``` lua
local on_exit = function(...)
    if c.get().on_exit then
        c.get().on_exit(...)
    end
    print ("exit")
    -- client = nil
    -- id = nil
end
```

resulting in no error on quitting Neovim.

Is there really any need to set client and id to nil here ?

Hope this explains this enough.

Cheers

### Requirements

- [X] I have read and followed the instructions above and understand that my issue will be closed if I did not provide the required information.

**Comment by jose-elias-alvarez** [comment]
> Is there really any need to set client and id to nil here ?

Yes, as otherwise we would maintain a reference to an inactive client after exiting.

I think I see the issue - we schedule sending notifications, so it's possible for a notification to be sent with a `nil` client ID. If I'm right, this is the fix I'd prefer:

```diff
diff --git a/lua/null-ls/client.lua b/lua/null-ls/client.lua
index 4a99224..9ee60a0 100644
--- a/lua/null-ls/client.lua
+++ b/lua/null-ls/client.lua
@@ -216,7 +216,8 @@ M.retry_add = function(bufnr)
 end
 
 M.send_progress_notification = function(token, opts)
-    if not id then
+    local client_id = id -- copy into closure to avoid race condition
+    if not client_id then
         return
     end
 
@@ -235,7 +236,7 @@ M.send_progress_notification = function(token, opts)
                 message = opts.message,
             },
         }, {
-            client_id = id,
+            client_id = client_id,
         })
     end)
 end
```

Could you try that out and see if it fixes your issue?

**Comment by thecontinium** [comment]
Yes - this fixes it. 
Thanks.

**Comment by jose-elias-alvarez** [comment]
Thanks for verifying - I merged the fix in 716dec7be9f92a516d968ad534548187384c4089.

**Comment by thecontinium** [comment]
Cheers
